services/remotecluster/send.go

```diff
@@ -0,0 +1,144 @@
+//
+// An optional callback can be provided that receives the success or fail result of sending to each remote cluster.
+// Success or fail is regarding message delivery only.  If a callback is provided it should return quickly.
+func (rcs *Service) SendOutgoingMsg(ctx context.Context, msg *model.RemoteClusterMsg, f SendResultFunc) error {
+	task := sendTask{
+		msg: msg,
+		f:   f,
+	}
+
+	if ctx == nil {
+		ctx = context.Background()
+	}
+
+	select {
+	case rcs.send <- task:
+		return nil
+	case <-ctx.Done():
+		return NewBufferFullError(cap(rcs.send))
+	}
+}
+
+func (rcs *Service) sendLoop(done chan struct{}) {
+	// create thread pool for concurrent message sending.
+	for i := 0; i < MaxConcurrentSends; i++ {
+		go func() {
+			for {
+				select {
+				case task := <-rcs.send:
+					if task.msg != nil {
+						rcs.sendMsg(task)
+					}
+				case <-done:
+					return
+				}
+			}
+		}()
+	}
+}
+
+func (rcs *Service) sendMsg(task sendTask) {
+	// get list of interested remotes.
+	list, err := rcs.server.GetStore().RemoteCluster().GetByTopic(task.msg.Topic)
+	if err != nil {
+		if task.f != nil {
+			task.f(task.msg, nil, nil, fmt.Errorf("could not send msg to remote: %w", err))
+		}
+	}
+
+	// bound the number of concurrent goroutines used to send using a semaphore.
+	bound := make(chan struct{}, MaxConcurrentSends)
+
+	for _, rc := range list {
+		bound <- struct{}{}
+		go func(rc *model.RemoteCluster) {
+			defer func() { <-bound }()
+			resp, err := rcs.sendMsgToRemote(rc, task)
+			if task.f != nil {
+				task.f(task.msg, rc, resp, err)
+			}
+			if err != nil {
+				rcs.server.GetLogger().Log(mlog.LvlRemoteClusterServiceError, "Remote Cluster message send failed",
+					mlog.String("remote", rc.DisplayName), mlog.String("msgId", task.msg.Id), mlog.Err(err))
+			} else {
+				rcs.server.GetLogger().Log(mlog.LvlRemoteClusterServiceDebug, "Remote Cluster message sent successfully",
+					mlog.String("remote", rc.DisplayName), mlog.String("msgId", task.msg.Id))
+			}
+		}(rc)
+	}
+}
+
+func (rcs *Service) sendMsgToRemote(rc *model.RemoteCluster, task sendTask) ([]byte, error) {
+	frame := &model.RemoteClusterFrame{
+		RemoteId: rc.RemoteId,
+		Token:    rc.RemoteToken,
+		Msg:      task.msg,
+	}
+	url := fmt.Sprintf("%s/%s", rc.SiteURL, SendMsgURL)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Millisecond*SendTimeoutMillis)
+	defer cancel()
+
+	return rcs.sendFrameToRemote(ctx, frame, url)
+}
+
+func (rcs *Service) sendFrameToRemote(ctx context.Context, frame *model.RemoteClusterFrame, url string) ([]byte, error) {
+	if ctx == nil {
+		ctx = context.Background()
+	}
+
+	body, err := json.Marshal(frame)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest("POST", url, bytes.NewReader(body))
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := http.DefaultClient.Do(req.WithContext(ctx))
```

Comment:
Good catch pointing that out.  I did look at the defaults when I wrote that and they look sane for this use case.  I probably should have added a comment to that effect.

Specifically I expect no more than two concurrent requests (sendMsg and Ping) to the same host, and anything more than that can block.  Keep-alive is on, and the idle timeout looks fine.  The only one I'm not 100% sure about is MaxIdleConns at 100.  Only an issue if 50+ remotes are added, and even then not necessarily a problem as it will just be more aggressive killing idle connections.

Reply:
Not really a fan of using the default client as it could be modified/used someplace else in the code and have an effect here. Even if we use the defaults, I think we should go for our own instance with `client := &http.Client{}`